staging/src/k8s.io/client-go/util/workqueue/queue.go

```diff
@@ -175,22 +181,69 @@ func (q *Type) Done(item interface{}) {
 	q.metrics.done(item)
 
 	q.processing.delete(item)
+	if q.processing.len() == 0 {
+		q.cond.Signal()
+	}
 	if q.dirty.has(item) {
 		q.queue = append(q.queue, item)
 		q.cond.Signal()
 	}
 }
 
+// isProcessing indicates if there are still items on the work queue being
+// processed. It's used to drain the work queue on an eventual shutdown.
+func (q *Type) isProcessing() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.processing.len() != 0
+}
+
+// shouldDrain indicates if the queue shutdown should drain.
+func (q *Type) shouldDrain() bool {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	return q.drain
+}
+
 // ShutDown will cause q to ignore all new items added to it. As soon as the
 // worker goroutines have drained the existing items in the queue, they will be
 // instructed to exit.
 func (q *Type) ShutDown() {
+	q.shutdown()
+	for q.isProcessing() && q.shouldDrain() {
+		q.wait()
+	}
+}
+
+// ShutDownNoDrain will cause q to ignore all new items added to it and
+// immediately instruct the worker goroutines to exit. It is safe to call
+// ShutDownNoDrain in combination with ShutDown, as to force the queue shut down
+// to terminate immediately, if needed.
+func (q *Type) ShutDownNoDrain() {
+	q.noDrain()
+	q.shutdown()
+}
+
+func (q *Type) wait() {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	q.cond.Wait()
+}
+
+func (q *Type) shutdown() {
 	q.cond.L.Lock()
 	defer q.cond.L.Unlock()
 	q.shuttingDown = true
 	q.cond.Broadcast()
 }
 
+func (q *Type) noDrain() {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+	q.drain = false
+	q.cond.Broadcast()
```

Comment:
I'm not sure both `noDrain` and `shutdown` need to broadcast?

Reply:
Indeed, given that they are called successively it is fine to only have `shutdown` broadcast.